libkbfs/tlf_edit_history.go

```diff
@@ -348,26 +434,168 @@ func (teh *TlfEditHistory) GetComplete(ctx context.Context,
 	if currEdits == nil {
 		// We broke out of the loop early.
 		var err error
-		currEdits, err = teh.calculateEditCounts(ctx, rmds)
+		currEdits, _, err = teh.calculateEditCounts(ctx, rmds)
 		if err != nil {
 			return nil, err
 		}
 	}
 
-	// Sort each of the edit lists by timestamp
-	for w, list := range currEdits {
-		sort.Sort(list)
-		if len(list) > desiredEditsPerWriter {
-			list = list[len(list)-desiredEditsPerWriter:]
+	teh.setEdits(ctx, currEdits, rmds)
+	return teh.getEditsCopyLocked(), nil
+}
+
+func (teh *TlfEditHistory) updateHistory(ctx context.Context,
+	rmds []ImmutableRootMetadata) error {
+	defer teh.wg.Done()
+	if len(rmds) == 0 {
+		return nil
+	}
+	teh.log.CDebugf(ctx, "Processing %d MDs for notifications "+
+		"(most recent revision: %d)", len(rmds), rmds[len(rmds)-1].Revision)
+
+	currEdits := teh.getEditsCopy()
+	if currEdits == nil {
+		teh.log.CDebugf(ctx, "No history to update; ignoring")
+		return nil
+	}
+
+	wasComplete := currEdits.isComplete()
+
+	newEdits, chains, err := teh.calculateEditCounts(ctx, rmds)
+	if err != nil {
+		return err
+	}
+
+	// Which paths have been removed?
+	removed := make(map[string]bool)
+	// TODO: can I used chains.deletedOriginals instead?  It's hard to
+	// get the full path that way.
+	for _, chain := range chains.byOriginal {
+		for _, op := range chain.ops {
+			rop, ok := op.(*rmOp)
+			if !ok {
+				continue
+			}
+			path := rop.getFinalPath().String() + "/" + rop.OldName
+			// A rename op might show later that this was only renamed.
+			removed[path] = true
 		}
-		currEdits[w] = list
 	}
-	teh.log.CDebugf(ctx, "Edits complete: %d revisions, starting from "+
-		"revision %d", len(rmds), rmds[0].Revision)
 
-	teh.lock.Lock()
-	defer teh.lock.Unlock()
-	teh.edits = currEdits
-	teh.editsTime = teh.config.Clock().Now()
-	return teh.getEditsCopyLocked(), nil
+	// Which paths have been renamed?
+	renamed := make(map[string]string)
+	for original, ri := range chains.renamedOriginals {
+		// Find both the old path and the new path using the old and
+		// new parents (which are each guaranteed to have at least one
+		// op, since the rename operation is split up into two ops).
+		oldParentChain, ok := chains.byOriginal[ri.originalOldParent]
+		if !ok || len(oldParentChain.ops) == 0 {
+			teh.log.CDebugf(ctx, "Couldn't find old parent to a rename "+
+				"op for original ptr %v", original)
+		}
+		oldPath := oldParentChain.ops[0].getFinalPath().String() + "/" +
+			ri.oldName
+
+		newParentChain, ok := chains.byOriginal[ri.originalNewParent]
+		if !ok || len(newParentChain.ops) == 0 {
+			teh.log.CDebugf(ctx, "Couldn't find new parent to a rename "+
+				"op for original ptr %v", original)
+		}
+		newPath := newParentChain.ops[0].getFinalPath().String() + "/" +
+			ri.newName
+
+		// Ignore any previous rmOps.
+		delete(removed, oldPath)
+		// If a file was overwritten, ignore all the old edits.
+		removed[newPath] = true
+		// Rename the file.
+		renamed[oldPath] = newPath
```

Comment:
`crChains` takes care of that magically for you, by eliminating all the interior renames.  So `chains.renamedOriginals` will only have the original start and final end name for any given file.